[MRG+1] Support for 64 bit sparse array indices in text vectorizers

rth opened the pull request:
#### Reference Issue

Continuation of https://github.com/scikit-learn/scikit-learn/pull/6473 (itself continuation of https://github.com/scikit-learn/scikit-learn/pull/6194). Fixes https://github.com/scikit-learn/scikit-learn/issues/6183 (in CountVectorizer); also fixes https://github.com/scikit-learn/scikit-learn/issues/8941 (in HashingVectorizer).

Closes #6473
Closes #6194
Closes #6183 
Closes #8941

#### What does this implement/fix? Explain your changes.

This PR switches to 64 bit indexing by default for `indptr` array indices in  `CountVectorizer`, `TfidfVectorizer` and `HashingVectorizer`. It follows the following assumptions,
 * `indptr` and `indices` attributes  have the same dtype in Scipy's CSR arrays [[1](https://github.com/scikit-learn/scikit-learn/pull/6194#issuecomment-176118626)]
 * when 64 bit `indptr` and `indices` are given to `csr_matrix` constructor, it will downcast them to 32 bit if this are sufficient to hold their contents [[2](https://github.com/scikit-learn/scikit-learn/pull/6473#issuecomment-309111537)]
 *  overflow happens in `indptr` which is typically negligible in size when compared to `indices`

As a result, in this PR,
  *  in the intermediary array representation `indptr` is always 64 bit, while `indices` is either without dtype (in `CountVectorizer`, etc) or 32 bit (in `HashingVectorizer` as the hash size if 32 bit anyway).
  * the returned arrays use 32 bit indexing for both `indptr` and `indices` if this dtype is sufficient, or use 64 bit indexing otherwise (if it's supported with scipy > 0.14).  Which makes this PR fully backward compatible.

#### Any other comments?

All the benchmark scripts and results can be found in [here](https://github.com/rth/notebooks/tree/master/sklearn/PR_6473).

* This PR was benchmarked on 20 newsgroups dataset with `bench_text_vectorizers.py` from https://github.com/scikit-learn/scikit-learn/pull/9086 : results before this PR (`bench_master.log`) and after this PR (`bench_PR.log`) show no degradation in performance or memory usage.

* The overflow from the original issues were reproduced with `run.sh` script, and on `master` the output is in `res_master.log`. After this PR, indices in  all  text vectors don't overflow `res_PR.log` and use 64 bit indexing when necessary. At least 90 GB RAM is required by `run.sh` (tested on EC2 `m4.10xlarge` with 160 GB RAM).

* ~~norm='l2' will still fail due to cython code (related to https://github.com/scikit-learn/scikit-learn/issues/2969)~~ (fixed in https://github.com/scikit-learn/scikit-learn/pull/9663)

* These changes were reviewed in https://github.com/scikit-learn/scikit-learn/pull/6194 by @lesteve @ogrisel . This PR extends them to `HashingVectorizer` and accounts for latest modifications  in vectorizers.

* ~~64 bit indexing  [won't work on Windows](https://github.com/scikit-learn/scikit-learn/pull/6194#issuecomment-176118626)~~

@psinger Would you be able to confirm that this fixes the overflow in CountVectorizer indexing? Thanks. 

cc @jnothman

Commit:
Support new scipy sparse array indices, which can now be > 2^31 (< 2^63).
This is needed for very large training sets.
Feature indices (based on the number of distinct features), are unlikely to need 4 bytes per value, however.

Commit:
Also increase size of integer values in indptr in the next step.

Commit:
Use long for both arrays if scipy >= 0.14.
Tweak comments

rth committed:
Reusing make_int_array function

jnothman commented:
I'd rather leave this until after release. ping then

On 18 Jun 2017 4:39 am, "Roman Yurchak" <notifications@github.com> wrote:

> Reference Issue
>
> Continuation of #6473
> <https://github.com/scikit-learn/scikit-learn/pull/6473> (itself
> continuation of #6194
> <https://github.com/scikit-learn/scikit-learn/pull/6194>). Fixes #6183
> <https://github.com/scikit-learn/scikit-learn/issues/6183> (in
> CountVectorizer); also fixes #8941
> <https://github.com/scikit-learn/scikit-learn/issues/8941> (in
> HashingVectorizer).
> What does this implement/fix? Explain your changes.
>
> This PR switches to 64 bit indexing by default for indptr array indices
> in CountVectorizer, TfidfVectorizer and HashingVectorizer. It follows the
> following assumptions,
>
>    - indptr and indices attributes have the same dtype in Scipy's CSR
>    arrays [1
>    <https://github.com/scikit-learn/scikit-learn/pull/6194#issuecomment-176118626>
>    ]
>    - when 64 bit indptr and indices are given to csr_matrix constructor,
>    it will downcast them to 32 bit if this are sufficient to hold their
>    contents [2
>    <https://github.com/scikit-learn/scikit-learn/pull/6473#issuecomment-309111537>
>    ]
>    - overflow happens in indptr which is typically negligible in size
>    when compared to indices
>
> As a result, in this PR,
>
>    - in the intermediary array representation indptr is always 64 bit,
>    while indices is either without dtype (in CountVectorizer, etc) or 32
>    bit (in HashingVectorizer as the hash size if 32 bit anyway).
>    - the returned arrays use 32 bit indexing for both indptr and indices
>    if this dtype is sufficient, or use 64 bit indexing otherwise (if it's
>    supported with scipy > 0.14). Which makes this PR fully backward compatible.
>
> Any other comments?
>
> All the benchmark scripts and results can be found in here
> <https://github.com/rth/notebooks/tree/master/sklearn/PR_6473>.
>
>    -
>
>    This PR was benchmarked on 20 newsgroups dataset with
>    bench_text_vectorizers.py from #9086
>    <https://github.com/scikit-learn/scikit-learn/pull/9086> : results
>    before this PR (bench_master.log) and after this PR (bench_PR.log)
>    show no degradation in performance or memory usage.
>    -
>
>    The overflow from the original issues were reproduced with run.sh
>    script, and on master the output is in res_master.log. After this PR,
>    indices in all text vectors don't overflow res_PR.log and use 64 bit
>    indexing when necessary. At least 90 GB RAM is required by run.sh
>    (tested on EC2 m4.10xlarge with 160 GB RAM).
>    -
>
>    norm='l2' will still fail due to cython code (related to #2969
>    <https://github.com/scikit-learn/scikit-learn/issues/2969>)
>    -
>
>    These changes were reviewed in #6194
>    <https://github.com/scikit-learn/scikit-learn/pull/6194> by @lesteve
>    <https://github.com/lesteve> @ogrisel <https://github.com/ogrisel> .
>    This PR extends them to HashingVectorizer and accounts for latest
>    modifications
>    in vectorizers.
>    -
>
>    64 bit indexing won't on Windows
>    <https://github.com/scikit-learn/scikit-learn/pull/6194#issuecomment-176118626>
>
> @psinger <https://github.com/psinger> Would you be able to confirm that
> this fixes the overflow in CountVectorizer indexing? Thanks.
>
> cc @jnothman <https://github.com/jnothman>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scikit-learn/scikit-learn/pull/9147
> Commit Summary
>
>    - Support new scipy sparse array indices, which can now be > 2^31 (<
>    2^63).
>    - Also increase size of integer values in indptr in the next step.
>    - Use long for both arrays if scipy >= 0.14.
>    - Reusing make_int_array function
>    - Extend changes to the HashingVectorizer
>
> File Changes
>
>    - *M* sklearn/feature_extraction/_hashing.pyx
>    <https://github.com/scikit-learn/scikit-learn/pull/9147/files#diff-0>
>    (13)
>    - *M* sklearn/feature_extraction/text.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9147/files#diff-1>
>    (29)
>
> Patch Links:
>
>    - https://github.com/scikit-learn/scikit-learn/pull/9147.patch
>    - https://github.com/scikit-learn/scikit-learn/pull/9147.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9147>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60Hy4MOjXSLoGXHfZJvmaLBXR4ycks5sFB1hgaJpZM4N9TOv>
> .
>

rth commented:
Sure, will do.

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -789,8 +791,20 @@ def _count_vocab(self, raw_documents, fixed_vocab):
+                indices_dtype = np.int_
+            else:
+                raise ValueError(('sparse CSR array has {} non-zero '
+                                  'elements and require 64 bit indexing, '
```
*requires

jnothman reviewed (commented):
I'm okay with this for what it is. My main concern here is interoperability. Producing a large sparse matrix, but then breaking when it is pushed into a downstream predictor that only supports int32-indexed data is not great. Thanks for your work at #2969 towards this. I'd like to fix the most critical of those before merging this, really.

rth commented:
Thank you for the review @jnothman ! I don't know how you manage to keep track of all the PRs and issues ) Sure makes sense to wait until predictors downstream are able to handle 64bit sparse arrays. Will try to make a few PR in that direction in the following weeks. 

Also Windows CI is currently failing in this PR because `long` is 32 bit on windows, I need to fix that...

jnothman commented:
> I don't know how you manage to keep track of all the PRs and issues

Uhh.... 
<img width="811" alt="screen shot 2017-08-14 at 6 05 35 pm" src="https://user-images.githubusercontent.com/78827/29263246-5f797766-811b-11e7-8457-a4dc4a2bd1c9.png">

I promise as much gets lost as gets tracked...

jnothman commented:
btw can we get windows support in python 3 using a long long array?

On 14 Aug 2017 6:05 pm, "Roman Yurchak" <notifications@github.com> wrote:

> Thank you for the review @jnothman <https://github.com/jnothman> ! I
> don't know how you manage to keep track of all the PRs and issues ) Sure
> makes sense to wait until predictors downstream are able to handle 64bit
> sparse arrays. Will try to make a few PR in that direction in the following
> weeks.
>
> Also Windows CI is currently failing in this PR because long is 32 bit on
> windows, I need to fix that...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9147#issuecomment-322125038>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xGBbuEnO3hsmq1eP_ysQ6hg_qeBks5sX_-zgaJpZM4N9TOv>
> .
>

rth commented:
> > I don't know how you manage to keep track of all the PRs and issues
>
> screen shot 2017-08-14 at 6 05 35 pm

I know, still, it's a lot of notifications per day...

> btw can we get windows support in python 3 using a long long array?

Was thinking along those lines as well, will try to.

rth committed:
Rewrite the 64 bit index support of CSR arrays

rth commented:
Following earlier discussion, I rewrote this PR somewhat. 

The current situation is that,
   * internally
       (a) for PY >3.3 on all platforms `longlong` indices (i.e. 64 bit) are going to be used for `indptr`
       (b) otherwise `long` indices are used for `indptr` (i.e. 64 bit on Unix like and 32 bit on Windows)
 The `indices` remain 32 bit in any case (i.e. the vocabulary size must be lower than 2e9) which allows to avoid memory overhead.

   * externally (i.e. from the users' perspective)
      (a) if either Linux (any Python version) or Win (with Python >3.3) is used, the resulting sparse arrays will have 32 bit or 64 bit indices as necessary. The 64 bit sparse array indices will only work for scipy >= 0.14, and error is raised otherwise.
      (b) if Windows is used with Python 2, only 32 bit indices are supported. A similar overflow error message would be printed to the one observed in the parent issues. 

I can't come up with a way to write unit tests for the 64bit case, however, I have updated [tests/benchmarks here](https://github.com/rth/ipynb/tree/master/sklearn/PR_9147) (on Linux) that test this PR on an actual dataset that produces 64 bit indices (and takes several hours to run). Codecov shows a decrease in coverage because there are no tests for the 64bit case.

>  My main concern here is interoperability. Producing a large sparse matrix, but then breaking when it is pushed into a downstream predictor that only supports int32-indexed data is not great. Thanks for your work at #2969 towards this. I'd like to fix the most critical of those before merging this, really.

With respect to this point, what currently works is normalization (fixed in https://github.com/scikit-learn/scikit-learn/pull/9663), dimensionality reduction (LSI, NMF), and a few linear models (e.g. LogisticRegression with lbfgs/cd solver & ElasticNet). I have not tested clustering yet. Of course, there is still work to do (I added the progress status to the parent comment of #2969), but this might be enough to have some reasonable workflow.  

A review would be appreciated, all tests now pass.

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -784,7 +785,8 @@ def _count_vocab(self, raw_documents, fixed_vocab):
         analyze = self.build_analyzer()
         j_indices = []
-        indptr = _make_int_array()
+        indptr = []
```
I checked that switching from an `array.array` to `list` doesn't have any negative performance impact here. `j_indices` is already a list and that's where most of the data is (since `j_indices` is much longer than `indptr`); `indptr` doesn't really matter. This simplifies things with respect to typing (and possible overflows) though.

jnothman commented:
> My main concern here is interoperability. Producing a large sparse matrix, but then breaking when it is pushed into a downstream predictor that only supports int32-indexed data is not great. 

Let's hurry in a fix for the liblinear segfault (#9545). Otherwise, yes, I'm happy to see this merged.

jnothman commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -79,4 +89,18 @@ def transform(raw_X, Py_ssize_t n_features, dtype, bint alternate_sign=1):
-    return (indices_a, np.frombuffer(indptr, dtype=np.int32), values[:size])
+    indptr_a = np.frombuffer(indptr, dtype=indices_np_dtype)
+
+    if indptr[-1] > 2147483648:  # = 2**31
```
It would sort of be nice if this were refactored somewhere, but I can't think of somewhere both pleasant and useful to keep it shared.

jnothman reviewed (commented):
This LGTM. Any chance @ogrisel could review?

glemaitre commented:
Code-wise LGTM.

Even if this is not linked to this PR, by reviewing, I found strange to exactly reallocate to specific size the indices at each iteration [there](https://github.com/scikit-learn/scikit-learn/pull/9147/files#diff-c39016cfecd02142fd271cbf6fbd6583R74). I would have think that it would be less costly to double the capacity and return a shrinked array.

rth committed:
Add what's new entry

rth committed:
Merge branch 'master' into large-number-of-features

rth commented:
Thanks for the review @jnothman and @glemaitre ! 

I added the corresponding what's new entry.

jnothman reviewed (commented):
Merging, thanks!

rth commented:
Thanks again for the review @jnothman and @glemaitre.

@mannby thank you for the initial implementation!

eric-wieser commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -33,12 +35,20 @@ def transform(raw_X, Py_ssize_t n_features, dtype, bint alternate_sign=1):
+    else:
+        # On Windows with PY2.7 long int would still correspond to 32 bit. 
+        indices_array_dtype = "l"
+        indices_np_dtype = np.int_
```
Can you just use `np.intp` for all cases here?

eric-wieser replied:
`np.dtype(np.intp).char` will give you that.
